Pulp tunings (#2565)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-pulpcore-api-workers.adoc

[id="Configuring_Pulpcore_API-Workers_{context}"]
= Configuring Pulpcore API Workers

> The pulpcore-api workers respond to incoming API requests.

Reviewer: And some content types, such as containers.

> Fewer API workers consume less memory and results in better performance.

Reviewer: This is questionable advice. Not enough workers means you can't serve the requests in parallel. Taking your advice to the extreme means I simply have 1 worker. But with enough clients, you can saturate that and performance is lower than it could be.

The API service is used by Foreman (via Katello), but also by container content. If your tests didn't test any container workloads then you're only testing a fraction of what it needs to serve.

Reviewer 2: @Imaanpreet Is this something you've observed? Do you have any numbers on this?

[id="Configuring_Pulpcore_API-Workers_{context}"]
= Configuring Pulpcore API Workers

> The pulpcore-api workers responds to incoming API requests. Less api workers consume less memory and results in better performance.

Reviewer 2: ```suggestion
The pulpcore-api workers respond to incoming API requests.
Fewer API workers consume less memory and results in better performance.
```

For example, we have compared these two setups:

[width="100%",cols="50%,50%",options="header",]
|===
|{Project} VM with 8 CPUs, 32 GiB RAM, 3 pulpcore-api_workers |{Project} VM with 8 CPUs, 32 GiB RAM, 10 pulpcore-api_workers
|===

> Outcome looks similar in both cases, so the recommendation would be to lower the number of api workers  because of the improvement in memory consumption.

Reviewer 2: ```suggestion
Outcome looks similar in both cases, so the recommendation would be to lower the number of API workers because of the improvement in memory consumption.
```

> .Use the below procedure to tune the pulpcore API workers:
> 

Reviewer 2: ```suggestion
.Procedure
```

> . Update the pulpcore api_service_worker_count on your {ProjectServer} or {SmartProxyServer} in the custom-hiera.yaml file.:

Reviewer 2: ```suggestion
. Update the pulpcore api_service_worker_count on your {ProjectServer} or {SmartProxyServer} in the `custom-hiera.yaml` file.:
```

I am unsure if we should recommend users editing this file vs. using foreman-installer.

Reviewer: I've always taken a stance that `custom-hiera.yaml` is unsupported. Anything that's written in docs, so we can't refer to it.

If there's always a recommendation to lower the number of workers, that should be filed as a bug. Telling users "we have poor defaults, always change this" is a poor experience.

My suggestion is to drop this chapter altogether.

Reviewer: And if it is common to change it, an RFE to add a parameter as was done with the worker count is also a valid solution.

Author: Hello @ekohl: I totally agree on `custom-hiera.yaml is unsupported` part. Do you mean to dropping off this section or proc_configuring-pulpcore-workers as well?

Reviewer: I'm questioning the usefulness of `proc_configuring-pulpcore-workers`. When I read it, it comes down to "here's a parameter you can change, but we noticed it doesn't make a difference". So as a reader of that, is it adding any value or just confusing me?

Author: cool, let's drop this section.

Reviewer: > Regarding `proc_configuring-pulpcore-api-workers.adoc` section, can we suggest updating pulpcore api_service_worker_count with another method, for example - via foreman-installer?

Currently this is the code that sets it:
https://github.com/theforeman/puppet-pulpcore/blob/297a48acafcbdb9871689b6e267598c7414692c2/manifests/init.pp#L246-L247

https://projects.theforeman.org/issues/36957 was opened by @maximiliankolb to expose a parameter for this and I just opened https://github.com/theforeman/puppet-foreman_proxy_content/pull/467.

I still think that most users shouldn't touch this and if our default tuning is poor, then it should be fixed.

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
pulpcore::api_service_worker_count: X
----

> +
> . Run the installer to apply your changes:
> +
> [options="nowrap", subs="+quotes,verbatim,attributes"]
> ----
> foreman-installer
> ----
> +
> 

Reviewer 2: ```suggestion
. Apply your changes:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# {foreman-installer}
----
```
